File: easy/3633_earliest_finish_time_for_land_and_water_ride_i/solution.py
from typing import List


class Solution:
    def earliestFinishTime(
        self,
        landStartTime: List[int],
        landDuration: List[int],
        waterStartTime: List[int],
        waterDuration: List[int],
    ) -> int:
        landTime = sorted(
            zip(landStartTime, landDuration), key=lambda t: t[0] + t[1]
        )
        waterTime = sorted(
            zip(waterStartTime, waterDuration), key=lambda t: t[0] + t[1]
        )

        earliestFinishTime = float("inf")

        landFt = landTime[0][0] + landTime[0][1]
        for waterSt, waterDur in waterTime:
            earliestFinishTime = min(
                earliestFinishTime, max(landFt, waterSt) + waterDur
            )

        waterFt = waterTime[0][0] + waterTime[0][1]
        for landSt, landDur in landTime:
            earliestFinishTime = min(
                earliestFinishTime, max(waterFt, landSt) + landDur
            )

        return earliestFinishTime

File: easy/3633_earliest_finish_time_for_land_and_water_ride_i/solution.cpp
#include <vector>

class Solution {
public:
  int earliestFinishTime(std::vector<int> &landStartTime,
                         std::vector<int> &landDuration,
                         std::vector<int> &waterStartTime,
                         std::vector<int> &waterDuration) {
    int earliestLandFinishTime = INT32_MAX;
    for (size_t i = 0; i < landStartTime.size(); ++i)
      earliestLandFinishTime =
          std::min(earliestLandFinishTime, landStartTime[i] + landDuration[i]);

    int earliestWaterFinishTime = INT32_MAX;
    for (size_t i = 0; i < waterStartTime.size(); ++i)
      earliestWaterFinishTime = std::min(earliestWaterFinishTime,
                                         waterStartTime[i] + waterDuration[i]);

    int earliestFinishTime = INT32_MAX;

    for (size_t i = 0; i < landStartTime.size(); ++i)
      earliestFinishTime =
          std::min(earliestFinishTime,
                   std::max(earliestWaterFinishTime, landStartTime[i]) +
                       landDuration[i]);
    for (size_t i = 0; i < waterStartTime.size(); ++i)
      earliestFinishTime =
          std::min(earliestFinishTime,
                   std::max(earliestLandFinishTime, waterStartTime[i]) +
                       waterDuration[i]);

    return earliestFinishTime;
  }
};
